File: btc_grocery/lib/controllers/review_controller.dart
import 'package:get/get.dart';
import '../../models/review_model.dart';
import '../../services/api/review_service.dart';

class ReviewController extends GetxController {
  final reviews = <Review>[].obs;
  final _service = ReviewService();

  Future<void> fetchReviews(int productId) async {
    try {
      final result = await _service.fetchReviews(productId);
      reviews.value = result;
    } catch (e) {
      Get.snackbar("Error", "Failed to fetch reviews: $e");
    }
  }

  Future<void> addReview(
    int userId,
    int productId,
    int rating,
    String comment,
  ) async {
    try {
      await _service.addReview(
        userId: userId,
        productId: productId,
        rating: rating,
        comment: comment,
      );
      fetchReviews(productId);
      Get.snackbar("Success", "Review added");
    } catch (e) {
      Get.snackbar("Error", "Could not add review: $e");
    }
  }

  Future<void> updateReview(
    int reviewId,
    int productId,
    int rating,
    String comment,
  ) async {
    try {
      await _service.updateReview(
        reviewId: reviewId,
        rating: rating,
        comment: comment,
      );
      fetchReviews(productId);
      Get.snackbar("Success", "Review updated");
    } catch (e) {
      Get.snackbar("Error", "Could not update review: $e");
    }
  }

  Future<void> deleteReview(int reviewId, int productId) async {
    try {
      await _service.deleteReview(reviewId);
      fetchReviews(productId);
      Get.snackbar("Success", "Review deleted");
    } catch (e) {
      print("Delete review error: $e");

      Get.snackbar("Error", "Could not delete review: ${e.toString()}");
    }
  }
}

File: Food-main/backend/controllers/reviewController.js
const Review = require('../models/Review');
const Food = require('../models/foodModel');
const Sequelize = require('../db');


/** 🔹 Get All Reviews for a Food */
exports.getReviewsByFood = async (req, res) => {
  const { foodId } = req.params;
  try {
    const reviews = await Review.findAll({ where: { foodId } });
    res.json(reviews);
  } catch (err) {
    console.log("Get Reviews Error:", err);
    res.status(500).json({ message: err.message });
  }
};

/** 🔹 Add a Review */
exports.addReview = async (req, res) => {
  const { userId, foodId, rating, comment } = req.body;
  try {
    const review = await Review.create({
      userId,
      foodId,
      rating,
      comment
    });
    res.json({ review, message: "Review added successfully" });
  } catch (err) {
    console.log("Add Review Error:", err);
    res.status(500).json({ message: err.message });
  }
};

/** 🔹 Update a Review */
exports.updateReview = async (req, res) => {
  const { id } = req.params;
  const { rating, comment } = req.body;
  try {
    const review = await Review.findByPk(id);
    if (!review) return res.status(404).json({ message: "Review not found" });

    review.rating = rating || review.rating;
    review.comment = comment || review.comment;
    await review.save();

    res.json({ review, message: "Review updated successfully" });
  } catch (err) {
    console.log("Update Review Error:", err);
    res.status(500).json({ message: err.message });
  }
};

/** 🔹 Delete a Review */
exports.deleteReview = async (req, res) => {
  const { id } = req.params;
  try {
    const review = await Review.findByPk(id);
    if (!review) return res.status(404).json({ message: "Review not found" });

    await review.destroy();
    res.json({ message: "Review deleted successfully" });
  } catch (err) {
    console.log("Delete Review Error:", err);
    res.status(500).json({ message: err.message });
  }
};
exports.getReviewsByRestaurant = async (req, res) => {
  const { restaurantId } = req.params;
  try {
    // Find all foods for that restaurant
    const foods = await Food.findAll({ where: { restaurantId } });
    const foodIds = foods.map(f => f.id);

    // Find reviews for those foods
    const reviews = await Review.findAll({
      where: { foodId: foodIds }
    });

    res.json(reviews);
  } catch (err) {
    console.log("Get Restaurant Reviews Error:", err);
    res.status(500).json({ message: err.message });
  }
};

/** 🔹 Get Average Rating for a Restaurant */
exports.getRestaurantAvgRating = async (req, res) => {
  const { restaurantId } = req.params;
  try {
    const foods = await Food.findAll({ where: { restaurantId } });
    const foodIds = foods.map(f => f.id);

    const result = await Review.findAll({
      attributes: [
        [Sequelize.fn('AVG', Sequelize.col('rating')), 'avgRating']
      ],
      where: { foodId: foodIds },
      raw: true
    });

    res.json({
      avgRating: parseFloat(result[0].avgRating || 0).toFixed(2)
    });
  } catch (err) {
    console.log("Get Restaurant Avg Rating Error:", err);
    res.status(500).json({ message: err.message });
  }
};